### PR TITLE
[ObservableProperty] generates partial OnPropertyChanging/Changed methods

### LINKED ISSUE #117: Discussion: Should the MVVM SourceGenerator generate partial methods for ObservableProperty fields
I wonder if it is useful when fields with the ObservableProperty attribute should also call a `On...Changing` and a `On...Changed` partial method. This would allow more extensibility. Look at this example code. The Customer has a name property. The source generator could generate two partial methods (OnNameChanging and OnNameChanged) which are called in the Setter of the property. When the methods are not implemented they call will automatically be removed by the compiler.

```CS
partial class Customer : ObservableObject {

    [ObservableProperty]
    private string _name;

}
```

The MVVM Source generator should generates this

```CS
partial class Customer : ObservableObject {

    [ObservableProperty]
    private string _name;

    [System.CodeDom.Compiler.GeneratedCode("CommunityToolkit.Mvvm.SourceGenerators.ObservablePropertyGenerator", "7.1.0.0")]
    [System.Diagnostics.DebuggerNonUserCode]
    [System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage]
    public string Name {
        get {
            return _name;
        }
        set {
            if (!System.Collections.Generic.EqualityComparer<string>.Default.Equals(_name, value)) {
                OnNameChanging(value);
                OnPropertyChanging(nameof(Name));
                _name = value;
                OnNameChanged(value);
                OnPropertyChanged(nameof(Name));
            }
        }
    }

    partial void OnNameChanging(string value);
    partial void OnNameChanged(string value);
}
```

The developer can now decide to put code in the partial method. For example some code to start some business logic.

```CS
partial class Customer  {

    partial void OnNameChanged(string value) {
        // Start some business logic
    }

}
```

The Roslyn compiler will merge the two partial class files into one. Because the OnNameChanging was not implemented it is removed.

```CS
class Customer : ObservableObject {

    [ObservableProperty]
    private string _name;

    [System.CodeDom.Compiler.GeneratedCode("CommunityToolkit.Mvvm.SourceGenerators.ObservablePropertyGenerator", "7.1.0.0")]
    [System.Diagnostics.DebuggerNonUserCode]
    [System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage]
    public string Name {
        get {
            return _name;
        }
        set {
            if (!System.Collections.Generic.EqualityComparer<string>.Default.Equals(_name, value)) {
                OnPropertyChanging(nameof(Name));
                _name = value;
                OnNameChanged(value);
                OnPropertyChanged(nameof(Name));
            }
        }
    }

    void OnNameChanged(string value) {
        // Start some business logic
    }
}
```

Maybe the parameter for the `On...Changing` method should be a `ref` parameter so it can be modified inside the implemented partial method. 

What do you think?

### PR DESCRIPTION
**Pull request by Sergio0694**
**Closes #117**

This PR generates `OnPropertyChanging` and `OnPropertyChanged` partial methods for `[ObservableProperty]`.

FYI @sonnemaf

### Example code

```csharp
namespace MyProject.ViewModels;

public partial class MyViewModel
{
    [ObservableProperty]
    private string? name;
}
```

### Generated code

```csharp
// <auto-generated/>
#pragma warning disable
#nullable enable
namespace MyProject.ViewModels
{
    partial class MyViewModel
    {
        [global::System.CodeDom.Compiler.GeneratedCode("CommunityToolkit.Mvvm.SourceGenerators.ObservablePropertyGenerator", "8.0.0.0")]
        [global::System.Diagnostics.DebuggerNonUserCode]
        [global::System.Diagnostics.CodeAnalysis.ExcludeFromCodeCoverage]
        public string? Name
        {
            get => name;
            set
            {
                if (!global::System.Collections.Generic.EqualityComparer<string?>.Default.Equals(name, value))
                {
                    OnNameChanging(value);
                    OnPropertyChanging(global::CommunityToolkit.Mvvm.ComponentModel.__Internals.__KnownINotifyPropertyChangingArgs.Name);
                    name = value;
                    OnNameChanged(value);
                    OnPropertyChanged(global::CommunityToolkit.Mvvm.ComponentModel.__Internals.__KnownINotifyPropertyChangedArgs.Name);
                }
            }
        }

        /// <summary>Executes the logic for when <see cref="Name"/> is changing.</summary>
        [global::System.CodeDom.Compiler.GeneratedCode("CommunityToolkit.Mvvm.SourceGenerators.ObservablePropertyGenerator", "8.0.0.0")]
        partial void OnNameChanging(string? value);
        /// <summary>Executes the logic for when <see cref="Name"/> just changed.</summary>
        [global::System.CodeDom.Compiler.GeneratedCode("CommunityToolkit.Mvvm.SourceGenerators.ObservablePropertyGenerator", "8.0.0.0")]
        partial void OnNameChanged(string? value);
    }
}
```

### CODE REVIEW

**Comment by michael-hawker** [comment]
@Sergio0694 in your comment above on the PR:

```cs
        /// <summary>Executes the logic for when <see cref="Name"/> is changing.</summary>
        [global::System.CodeDom.Compiler.GeneratedCode("CommunityToolkit.Mvvm.SourceGenerators.ICommandGenerator", "8.0.0.0")]
        
```

Should it be IPropertyChangeGenerator vs. `ICommandGenerator`? I guess I'm not sure what this generatedcode attribute is for?

File: tests/CommunityToolkit.Mvvm.UnitTests/Test_ObservablePropertyAttribute.cs [hunk 1]
```diff
@@ -273,6 +273,21 @@ public void Test_AlsoNotifyChangeFor()
         CollectionAssert.AreEqual(new[] { nameof(model.Surname), nameof(model.FullName), nameof(model.Alias) }, propertyNames);
     }
 
+    [TestMethod]
+    public void Test_OnPropertyChangingAndChangedPartialMethods()
```
Comment (michael-hawker) [inline tests/CommunityToolkit.Mvvm.UnitTests/Test_ObservablePropertyAttribute.cs:277]: Do we have tests somewhere that check that the OnChanging/OnChanged events fire in the right order and at the right time as setting the property values?

If so, do we want to replicate those here for these methods as well? Just guard against accidentally modifying behavior if code moves/refactors in the future?

**Review by michael-hawker — APPROVED** [review]
Looks good, just a suggestion for another test if you think it'd be good.

File: tests/CommunityToolkit.Mvvm.UnitTests/Test_ObservablePropertyAttribute.cs (hunk 1 above)
Comment (Sergio0694) [reply tests/CommunityToolkit.Mvvm.UnitTests/Test_ObservablePropertyAttribute.cs:277]: Good idea! Done in 17c2cf4725d1d20b08e326c358927870f8f4071f, thanks! 😄

**Review by Arlodotexe — APPROVED** [review]

**Pull request merged**